cns/fsnotify/fsnotify.go

```diff
@@ -55,29 +59,38 @@ func (w *watcher) releaseAll(ctx context.Context) {
 	w.lock.Lock()
 	defer w.lock.Unlock()
 	for containerID := range w.pendingDelete {
+		logger.Printf("deleting Endpoint asynchronously")
 		// read file contents
 		filepath := w.path + "/" + containerID
 		file, err := os.Open(filepath)
 		if err != nil {
-			w.log.Error("failed to open file", zap.Error(err))
+			logger.Errorf("failed to open file %s", err.Error())
 		}
 
 		data, errReadingFile := io.ReadAll(file)
 		if errReadingFile != nil {
-			w.log.Error("failed to read file content", zap.Error(errReadingFile))
+			logger.Errorf("failed to read file content %s", errReadingFile)
 		}
 		file.Close()
 		podInterfaceID := string(data)
-
-		w.log.Info("releasing IP for missed delete", zap.String("podInterfaceID", podInterfaceID), zap.String("containerID", containerID))
+		// in case of stateless CNI for Windows, CNS needs to remove HNS endpoitns first
+		if isStalessCNIWindows(w.cnsconfig) {
```

Comment:
The asyncdelete watcher should not need to know about the CNI mode. You need to compose it in such a way that it does the right thing with no additional knowledge instead of teaching it about all this extra CNS config and adding branches and complication.
One way you might accomplish this is by decorating the `ReleaseIPs` implementation that gets passed in during initialization. In `main` when we bulid this object, you already have all the context you need about if this is stateless CNI. _There_, wrap `ReleaseIPs` with this extra functionality and the call the underlying impl after you deleteEndpoint, like middleware.
I think you could implement that in a way that there are actually no changes to the fsnotify package at all, and only a small anonymous wrapper func added to this initialization in main.

Reply:
That was my first thought implementing this then I got some issues with the pattern of going through Rest CNS client and APIs internally from CNS. The async delete is not a direct request and I don't think we should go through Rest APIs for its related operations. For the ReleaseIP and GetEndpointState, I think we don't have any other choice other than CNS client and APIs since they are written that way. However, for HNS we still can avoid this.
I had some concerns about having let's say 250 async delete ops and as a result it will create 250*3 rest APIs for (releaseIP + GetEndpointState + DeleteHNSEndpoint). 
I'm more inclined to a solution to avoid going through cnsClient for this. Not sure if we can achieve it without tying up asynchdelte watched to extra packages.

Reply:
@tamilmani1989 Let me know what you think.